### PR TITLE
Restricting user access before login

### LINKED ISSUE #49: Require auth for all kinds of activity on the chat app
In order to keep a consistent story, we should require that all users should sign into the app before being able to view, create, or participate in chats. Unauthenticated users should be redirected to the login page.

As a stretch goal, we should have our landing page (the one we redirect unauthenticated users to) be a simple marketing page that explains what our app does, with a "Log in" link that takes them to the current login page. As an example, open https://hangouts.google.com/ in an incognito browser tab.

### PR DESCRIPTION
**Pull request by vocatwk**
I was working on the landing page and wanted to ask you guys if you agreed on the direction I was heading. Right now, the index.jsp file that users are redirected to, cuts off user access to the app before login and instead now gives a link to login page to users not logged in. If users are logged in(this will happen when logged in users click on "Codeu ChatApp - Nemo" or when they access the app after logging in),it displays the nav bar and so access to other areas of the app as well as links to conversation page and about page(third screen shot). 

Also, right now when you login you get redirected to conversation page but we can always have it redirect to the home page(third screenshot). Though i'm not sure what useful information the home page will display. 

There is not way to log out right now, but i'll fix that and when I do I was thinking it redirects to index.jsp which would then diplay a login link as if it were a new user. 

I'll work on this PR a little more before I merged, just wanted to know what you guys thought so far.

@yebrahim I guess this makes index.jsp the landing page, though i'm not sure if that's what you had in mind. Should I have created a new landing page and redirected users to the index page if they were logged in, istead of doing it in the same file?

screen shot 2018-06-11 at 8 47 13 am
screen shot 2018-06-11 at 8 47 21 am
screen shot 2018-06-11 at 8 47 52 am

resolves #49 #56

### CODE REVIEW

**Comment by lealgamb** [comment]
Also, this is for the future. The landing page can include what features our app has in order to convince users to join. We can based it off the Read Me file once it gets updated.

**Comment by yebrahim** [comment]
The screenshots here capture what I had in mind in the other discussion, so I'm on board with where you're headed. (I haven't reviewed the code here yet, will do once it's ready)

@lealgamb both are fine approaches, it's up to us to decide what we want to go with. A dedicated landing page seemed like a good spot to be creative about marketing our app (see for example https://www.google.com/photos/about/), but we're free to combine it with the login/register page of course.

**Comment by jasminechau2** [comment]
I really like the work you have done so far! It looks good. I think to make the login page as the landing page and adding a description to it

**Comment by lealgamb** [comment]
After looking at the sample landing page I definitely see the advantages of having a dedicated page to show off the features.

**Comment by yebrahim** [comment]
Is this ready for review?

**Comment by vocatwk** [comment]
Not yet, I was planning on adding css components to it so it could look better but should we talk about the design of the entire app before that? And I also haven't looked into security for it yet. I have a lot of time tonight so I will make sure it's ready before tomorrow as well as work on some of the other open issues.

**Comment by yebrahim** [comment]
You don't have to do everything in one go. If this is a complete piece and isn't breaking anything, we can get it reviewed and merged, while you think about making it better in a follow up PR.

**Comment by vocatwk** [comment]
Oh in that case then I guess it's ready for review. Sorry for the wait.

**Comment by vocatwk** [comment]
Yes, I meant to restrict on all pages. I got caught up on the landing page that I forgot. I'll push updates to this later.

**Comment by yebrahim** [comment]
Is there a way to catch all request paths and kick in the login code if the user isn't logged in? I don't know how we can do this, perhaps using a [filter](http://www.avajava.com/tutorials/lessons/what-is-a-filter-and-how-do-i-use-it.html), or something similar.

I'm concerned about adding the check code in every servlet since 1) this code can break at some point without us noticing, and 2) team members adding a new servlet can easily forget to add this code.

**Comment by vocatwk** [comment]
You're right. I had tried to do something similar to avoid all the repeat code as well but I didn't know about filters, I'll look into them and see if they can do the trick or try find something else like filters. Thanks!

**Comment by vocatwk** [comment]
it looks like filters are perfect for this, I'll revert the changes and use filters.

File: src/main/java/codeu/controller/SessionFilter.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codeu.controller;
+
+import codeu.model.data.User;
+import codeu.model.store.basic.UserStore;
+import codeu.model.store.basic.EventStore;
```
Comment (lealgamb) [inline src/main/java/codeu/controller/SessionFilter.java]: Doesn't seem like this is being used.

File: src/main/java/codeu/controller/SessionFilter.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    
+    if(username == null || userStore.getUser(username) == null) {
+      // user not logged in or not found
+      if(!(requestUrl.equals("/") || requestUrl.equals("/login") || requestUrl.equals("/register"))) {
+        //user does not have access, redirect to index.
+        request.getRequestDispatcher("/").forward(request,response);
+        return;
+      }
+    }
+    else if (requestUrl.equals("/login") || requestUrl.equals("/register")) {
```
Comment (lealgamb) [inline src/main/java/codeu/controller/SessionFilter.java]: nice feature 👍

File: src/main/java/codeu/controller/SessionFilter.java (hunk 1 above)
Comment (vocatwk) [reply src/main/java/codeu/controller/SessionFilter.java]: oops, will get rid of that.

**Comment by vocatwk** [comment]
Okay, I'll remove them. I created an issue for the tests and was going to do it in a different PR but I can push them to this PR as well.

**Comment by yebrahim** [comment]
Tests in a separate PR are totally fine. Just make sure you mention that in the PR description so that I don't nag you about it, and so others know where to find them when they're looking at this PR later on. :)

**Comment by vocatwk** [comment]
I initially pulled from master when I was about to push new changes but due to the errors I was getting(it was difficult to figure out), I reset head to see what the changes were using git diff and which commit had caused the errors. I added the changes back but this might've led to Jasmine's changes being included in this PR. Sorry!

**Comment by yebrahim** [comment]
You should be able to rebase to get just your changes in this branch. Do `git fetch` to make sure you have the latest changes, then do `git rebase origin/master` while you're on this branch.
What this does is, first it rolls back all of your changes all the way to where you branched off of master, updates the branch to match master, then re-applies your commits one by one on top. If any commit conflicts (because the base is now different, after updating the branch to match master), the rebase will stop, and ask you to resolve the conflicts manually. You can do that, then do `git add -A` and `git rebase --continue`. Repeat this until you have all of the changes resolved and committed. After that you can force push to this remote branch `git push -f`. You should now see only your changes in the PR.

**Comment by yebrahim** [comment]
I recommend doing this before merging to keep the PR clean, so that anyone looking at the history later on is not confused what it's trying to do. It should also make reverts cleaner. Let me know if you hit any issues with the rebase. It can get tricky, so any time you get stuck, you can abort the rebase by doing `git rebase --abort`.

**Comment by yebrahim** [comment]
After merging this, can you modify the README to add a sentence or two about the fact that we now filter user access on auth?

**Comment by vocatwk** [comment]
Yes, I'll make sure to update README. However, I thought I had jasmine's changes on this PR as well at first but under the files changed tab I can only see mine. Should I still do the rebase anyway?

**Comment by yebrahim** [comment]
The changes in `PersistentDataStore.java` and `navbar.jsp` are from her branch. I'd try the rebase anyway, if it becomes too complicated maybe it's not worth the effort. I'll let you decide.
I'll approve these changes just in case you wanted to merge so I don't block you.

**Comment by yebrahim** [comment]
The changes in `PersistentDataStore.java` and `navbar.jsp` are from her branch. I'd try the rebase anyway, if it becomes too complicated maybe it's not worth the effort. I'll let you decide.
I'll approve these changes just in case you wanted to merge so I don't block you.

**Comment by vocatwk** [comment]
I had to make changes in persistentDataStore.java and navBar.java as well just to fix incompatible changes made by me and her. However, my third commit still has all her other changes as well just because I did git reset HEAD~ after I pulled from master to figure out which commit was giving me errors and then I added some changes to fix those and committed the changes all back, so maybe the rebase is still needed?

**Comment by vocatwk** [comment]
I rebased it just to make sure and it went smoothly I think!

**Pull request merged**